battle-service/src/main/java/by/ibondarau/tictactoe/battleservice/service/impl/BattleServiceImpl.java

```diff
@@ -0,0 +1,176 @@
+package by.ibondarau.tictactoe.battleservice.service.impl;
+
+import by.ibondarau.tictactoe.battleservice.dao.BattleDao;
+import by.ibondarau.tictactoe.battleservice.exception.BusinessException;
+import by.ibondarau.tictactoe.battleservice.exception.NotFoundException;
+import by.ibondarau.tictactoe.battleservice.model.Battle;
+import by.ibondarau.tictactoe.battleservice.model.BattleResult;
+import by.ibondarau.tictactoe.battleservice.model.BattleStatus;
+import by.ibondarau.tictactoe.battleservice.model.FirstMoveRule;
+import by.ibondarau.tictactoe.battleservice.model.Move;
+import by.ibondarau.tictactoe.battleservice.service.BattleService;
+import by.ibondarau.tictactoe.battleservice.util.BattleUtils;
+import by.ibondarau.tictactoe.battleservice.util.RandomUtils;
+import com.sun.istack.NotNull;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.domain.Sort;
+import org.springframework.stereotype.Service;
+
+import java.time.ZonedDateTime;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+@Service
+public class BattleServiceImpl implements BattleService {
+
+    private final BattleDao battleDao;
+    private final RandomUtils randomUtils;
+
+    private final BattleUtils battleUtils;
+
+    @Autowired
+    public BattleServiceImpl(BattleDao battleDao, RandomUtils randomUtils, BattleUtils battleUtils) {
+        this.battleDao = battleDao;
+        this.randomUtils = randomUtils;
+        this.battleUtils = battleUtils;
+    }
+
+    @Override
+    public Battle createBattle(int playerId, int size, @NotNull FirstMoveRule firstMoveRule) {
+        Battle battle = new Battle()
+                .setFirstPlayerId(playerId)
+                .setSize(size)
+                .setFirstMoveRule(firstMoveRule)
+                .setCreated(ZonedDateTime.now().toInstant())
+                .setStatus(BattleStatus.CREATED);
+        battleDao.save(battle);
+        return battle;
+    }
+
+    @Override
+    public Battle getBattle(int battleId) {
+        return battleDao
+                .findBattleByBattleId(battleId)
+                .orElseThrow(() -> new NotFoundException("Battle with id " + battleId + " not found"));
+    }
+
+    @Override
+    public Battle joinBattle(int battleId, int secondPlayerId) {
+        Battle battle = battleDao.findBattleByBattleId(battleId).orElseThrow(() -> new NotFoundException("Battle with id " + battleId + " not found"));
+
+        if (!BattleStatus.CREATED.equals(battle.getStatus()) || battle.getSecondPlayerId() != null) {
+            throw new BusinessException("Unable to join: invalid battle status");
+        }
+
+        battle.setSecondPlayerId(secondPlayerId);
+        battle.setStatus(BattleStatus.STARTED);
+        battle.setStarted(ZonedDateTime.now().toInstant());
```

Comment:
i would assume that you can remove this field and rely on createdAt field, which could be implemented out of box with spring.